executor/index_lookup_join.go

```diff
@@ -360,49 +366,59 @@ func (ow *outerWorker) pushToChan(ctx context.Context, task *lookUpJoinTask, dst
 // When err is not nil, task must not be nil to send the error to the main thread via task.
 func (ow *outerWorker) buildTask(ctx context.Context) (*lookUpJoinTask, error) {
 	task := &lookUpJoinTask{
-		doneCh:            make(chan error, 1),
-		outerResult:       newFirstChunk(ow.executor),
-		encodedLookUpKeys: chunk.NewChunkWithCapacity([]*types.FieldType{types.NewFieldType(mysql.TypeBlob)}, ow.ctx.GetSessionVars().MaxChunkSize),
-		lookupMap:         mvmap.NewMVMap(),
+		doneCh:      make(chan error, 1),
+		outerResult: newList(ow.executor),
+		lookupMap:   mvmap.NewMVMap(),
 	}
 	task.memTracker = memory.NewTracker(stringutil.MemoizeStr(func() string { return fmt.Sprintf("lookup join task %p", task) }), -1)
+	task.outerResult.GetMemTracker().AttachTo(task.memTracker)
 	task.memTracker.AttachTo(ow.parentMemTracker)
 
 	ow.increaseBatchSize()
-	if ow.lookup.isOuterJoin { // if is outerJoin, push the requiredRows down
-		requiredRows := int(atomic.LoadInt64(&ow.lookup.requiredRows))
-		task.outerResult.SetRequiredRows(requiredRows, ow.maxBatchSize)
-	} else {
-		task.outerResult.SetRequiredRows(ow.batchSize, ow.maxBatchSize)
-	}
-
-	task.memTracker.Consume(task.outerResult.MemoryUsage())
-	for !task.outerResult.IsFull() {
-		err := Next(ctx, ow.executor, ow.executorChk)
+	requiredRows := ow.batchSize
+	if ow.lookup.isOuterJoin {
+		// If it is outerJoin, push the requiredRows down.
+		// Note: buildTask is triggered when `Open` is called, but
+		// ow.lookup.requiredRows is set when `Next` is called. Thus we check
+		// whether it's 0 here.
+		if parentRequired := int(atomic.LoadInt64(&ow.lookup.requiredRows)); parentRequired != 0 {
+			requiredRows = parentRequired
+		}
+	}
+	maxChunkSize := ow.ctx.GetSessionVars().MaxChunkSize
+	for requiredRows > task.outerResult.Len() {
+		chk := chunk.NewChunkWithCapacity(ow.outerCtx.rowTypes, maxChunkSize)
+		chk = chk.SetRequiredRows(requiredRows, maxChunkSize)
+		err := Next(ctx, ow.executor, chk)
 		if err != nil {
 			return task, err
 		}
-		if ow.executorChk.NumRows() == 0 {
+		if chk.NumRows() == 0 {
 			break
 		}
 
-		oldMemUsage := task.outerResult.MemoryUsage()
-		task.outerResult.Append(ow.executorChk, 0, ow.executorChk.NumRows())
-		newMemUsage := task.outerResult.MemoryUsage()
-		task.memTracker.Consume(newMemUsage - oldMemUsage)
+		task.outerResult.Add(chk)
```

Comment:
Update the `memTacker` here?

Reply:
In `Add(chk)`, it has updated the `memTacker`